src/bam/simulate_reads.rs

```diff
@@ -0,0 +1,165 @@
+    header.push_record(
+        bam::header::HeaderRecord::new(b"SQ")
+            .push_tag(b"SN", &ref_id)
+            .push_tag(b"LN", &(end - start)),
+    );
+    let mut bam_writer = bam::Writer::from_path(output_bam, &header, bam::Format::BAM)?;
+    for result in bam_reader.records() {
+        let mut record = result?;
+        if (record.cigar().end_pos() <= (end as i64))
+            && (record.mtid() == record.tid())
+            && (record.mpos() + 1 >= (start as i64))
+            && (record.mpos() + 1 < (end as i64))
+        //TODO Check if mate is fully in interval
+        {
+            record.cache_cigar();
+            //Check if mate record end within region
+            let artificial_seq = if record.is_unmapped() {
+                let mut seq = Vec::new();
+                add_random_bases(record.seq_len() as u64, &mut seq)?;
+                seq
+            } else {
+                build_sequence(
+                    &reference,
+                    &artificial_reference,
+                    &record,
+                    (start - 1) as usize,
+                )?
+            };
+            let artificial_record = build_record(&record, &artificial_seq, (start - 1) as i64)?;
+            bam_writer.write(&artificial_record)?;
+        }
+    }
+    Ok(())
+}
+
+fn build_record(record: &bam::Record, artificial_seq: &[u8], offset: i64) -> Result<bam::Record> {
+    let mut artificial_record = bam::record::Record::new();
+    if let Some(mate_cigar) = record.aux(b"MC") {
+        artificial_record.push_aux(b"MC", &mate_cigar);
+    }
+    artificial_record.set(
+        record.qname(),
+        Some(&record.cigar()),
+        artificial_seq,
+        record.qual(),
+    );
+    artificial_record.set_pos(record.pos() - offset);
+    artificial_record.set_tid(0);
+    artificial_record.set_mtid(0);
+    artificial_record.set_mpos(record.mpos() - offset);
+    artificial_record.set_flags(record.flags());
+    artificial_record.set_insert_size(record.insert_size());
+    artificial_record.set_mapq(record.mapq());
+    Ok(artificial_record)
+}
+fn build_sequence(
+    reference: &[u8],
+    artificial_reference: &[u8],
+    record: &bam::Record,
+    offset: usize,
+) -> Result<Vec<u8>> {
+    let mut artificial_seq = Vec::new();
+    let record_seq = record.seq().as_bytes();
+    let mut record_pos = 0;
+    let mut ref_pos = record.pos() as usize - offset;
+    //Create random seq for leading softclips
+    for cigar in record.cigar_cached().unwrap().iter() {
+        match cigar.char() {
+            'S' => {
+                add_random_bases(cigar.len() as u64, &mut artificial_seq)?;
+                record_pos += cigar.len() as usize;
+            }
+            'M' => {
+                (0..cigar.len()).for_each(|_| {
+                    let ref_base = artificial_reference.get(ref_pos).unwrap();
+                    if record_seq.get(record_pos).unwrap() == reference.get(ref_pos).unwrap() {
```

Comment:
You should reuse ref_base here.

Reply:
`ref_base` will be initialized [here](https://github.com/rust-bio/rust-bio-tools/blob/f4172657557d76cfc63c8aa3251bad7b2b597af8/src/bam/simulate_reads.rs#L109), now.
But the preinitialized value will never be used but overwritten the first time the corresponding matching arm is called.
I guess rust does not allow to init empty variables by design?

Reply:
it has the Option type for that purpose.

Reply:
Okay, it would be possible circumventing all the unwrapping. But than Option it is.